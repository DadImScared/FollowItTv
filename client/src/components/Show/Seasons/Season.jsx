
import React, { Component } from 'react';
import { connect } from 'react-redux';

import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Divider from '@material-ui/core/Divider';

import { getSeasonEpisodes, normalizeSeasonEpisodes } from '../../../actions/seasons';
import { makeSeasonEpisodes } from '../../../reducers/seasons';
import Episode from '../../Episode';


export class Season extends Component {

  async componentDidMount() {
    const { season: { id }, hasData, addSeasonEpisodes, addEpisodes, showId } = this.props;
    if (hasData) {
      return;
    }
    try {
      const { data } = await getSeasonEpisodes(id);
      const { episodeList, newEpisodes } = normalizeSeasonEpisodes(showId, data);
      addEpisodes(newEpisodes);
      addSeasonEpisodes(id, episodeList);
    }
    catch (e) {
      console.log(e);
    }
  }

  render() {
    const { season, episodes, show } = this.props;
    // const fullEpisodes = seasonEpisodes[season.id] || [];
    return (
      <div>
        <div style={{ marginBottom: '8px' }}>
          <div>
            <Typography gutterBottom>
              episodes: {season.episodeOrder}
            </Typography>
          </div>
          <Typography gutterBottom>
            {
              season.summary ?
                <span>summary: {season.summary}</span>
                :
                'No summary'
            }
          </Typography>
          <Divider/>
        </div>
        <Grid container>
          {
            episodes.map((episode, index) => {
              // const episode = episodes[episodeId] || {};
              return (
                <Grid item xs={12} lg={6} xl={4} style={{ marginBottom: '5px' }} key={`${episode.id}-${index}`}>
                  <Episode showActions={false} item={episode} show={show} />
                </Grid>
              );
            })
          }
        </Grid>
      </div>
    );
  }
}

const makeMapStateToProps = () => {
  const getSeasonEpisodesData = makeSeasonEpisodes();
  return (state, props) => {
    const results = getSeasonEpisodesData(state, props);
    console.log(getSeasonEpisodesData.recomputations());
    return results;
  };
};

export default connect(makeMapStateToProps)(Season);
