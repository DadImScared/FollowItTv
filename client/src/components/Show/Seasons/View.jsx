
import React from 'react';

import ExpansionPanel, {
  ExpansionPanelDetails,
  ExpansionPanelSummary
} from 'material-ui/ExpansionPanel';
import Typography from 'material-ui/Typography';
import ExpandMoreIcon from 'material-ui-icons/ExpandMore';

import Season from './Season';


const View = ({ seasons, changePanel, expanded, ...other }) => (
  <div>
    {
      Object.values(seasons).map((season, index) => (
        <ExpansionPanel
          expanded={expanded === index}
          onChange={(event, expanded) => changePanel(expanded, index)}
          key={`${season.id}-${index}`}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <Typography style={{ marginRight: '5px' }}>season {index + 1}</Typography>
            <Typography>aired {season.premiereDate} to {season.endDate}</Typography>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails>
            {
              expanded === index ?
                <Season season={season} {...other} />
                :
                <Typography>season {index + 1}</Typography>
            }
          </ExpansionPanelDetails>
        </ExpansionPanel>
      ))
    }
  </div>
);

export default View;
