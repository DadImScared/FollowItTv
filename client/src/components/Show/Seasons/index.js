
import React, { Component } from 'react';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { addSeasons, getSeasons, addSeasonEpisodes } from '../../../actions/seasons';
import { addEpisodes } from '../../../actions/episodes';
import View from './View';


export class Seasons extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      expanded: false
    };
  }

  componentDidMount() {
    if (this.props.show.id) {
      this.shouldGetSeasons(this.props.show.id);
    }
  }

  componentWillReceiveProps(nextProps) {
    const { show: oldShow } = this.props;
    const { show } = nextProps;
    if (show !== oldShow) {
      this.shouldGetSeasons(show.id);
    }
  }

  changePanel = (expanded, panel) => {
    this.setState({
      expanded: expanded ? panel:false
    });
  };

  shouldGetSeasons = async (showId) => {
    const { byId, addSeasons } = this.props;
    if (!byId[showId]) {
      const { data } = await getSeasons(showId);
      addSeasons(showId, data);
    }
  };

  render() {
    const { show = {}, byId, ...other } = this.props;
    return (
      <View {...this.state} {...other} show={show} changePanel={this.changePanel} seasons={byId[show && show.id] || {}} />
    );
  }
}

const mapStateToProps = ({ seasons: { byId } }) => ({
  byId
});

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    addSeasons,
    addEpisodes,
    addSeasonEpisodes
  }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(Seasons);
