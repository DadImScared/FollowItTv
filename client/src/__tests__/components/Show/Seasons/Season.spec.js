
import React from 'react';

import { shallow } from 'enzyme';

import { Season } from '../../../../components/Show/Seasons/Season';

describe('Season', () => {
  it('should render', () => {
    const props = {
      addEpisodes: jest.fn(),
      season: {
        id: 1,
        episodeOrder: 13
      },
      show: {},
      episodes: [{ id: 1 }],
      hasData: true,
      isLoading: false
    };
    const wrapper = shallow(<Season {...props} />);
    expect(props.addEpisodes).toHaveBeenCalledTimes(0);
    expect(wrapper).toMatchSnapshot();
  });
});
