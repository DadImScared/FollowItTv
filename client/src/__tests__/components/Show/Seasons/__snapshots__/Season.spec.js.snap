// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Season should render 1`] = `
<div>
  <div
    style={
      Object {
        "marginBottom": "8px",
      }
    }
  >
    <div>
      <WithStyles(Typography)
        gutterBottom={true}
      >
        episodes: 
        13
      </WithStyles(Typography)>
    </div>
    <WithStyles(Typography)
      gutterBottom={true}
    >
      No summary
    </WithStyles(Typography)>
    <WithStyles(Divider) />
  </div>
  <WithStyles(Grid)
    container={true}
  >
    <WithStyles(Grid)
      item={true}
      key="1-0"
      lg={6}
      style={
        Object {
          "marginBottom": "5px",
        }
      }
      xl={4}
      xs={12}
    >
      <WithStyles(Episode)
        item={
          Object {
            "id": 1,
          }
        }
        show={Object {}}
        showActions={false}
      />
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</div>
`;
